config/config.go

```diff
@@ -112,12 +119,33 @@ func (config *Config) CheckDiverseCombinations() error {
 	}
 	return nil
 }
+
+func pullFromUpstream(upstream Upstream, upstreamConf *UpstreamConf) ([]Server, error) {
+	servers, err := upstream.GetServers()
+	if err != nil {
+		return nil, err
+	}
+	for i := range servers {
+		servers[i].UpstreamConf = upstreamConf
+	}
+	return servers, nil
+}
+
 func parseUpstreams(config *Config) (err error) {
-	logged := false
+	var wg sync.WaitGroup
```

Comment:
Can we control the number of concurrency?

Reply:
I actually thought about this when implementing this. It turns out that even with ~40 upstreams, there's no significant memory usage spike. I don't think there's a need for limiting concurrency.